### ISSUE TITLE
Client component in App dir warning about use of window or localStorage

### ISSUE THREAD

**Issue by The-Code-Monkey** [vercel/next.js#44446]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:08:47 PST 2022; root:xnu-8792.61.2~4/RELEASE_X86_64
    Binaries:
      Node: 16.15.0
      npm: 8.5.5
      Yarn: 1.22.15
      pnpm: 6.11.0
    Relevant packages:
      next: 13.1.1
      eslint-config-next: 13.1.1
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

https://github.com/The-Code-Monkey/PaperCMS/blob/main/app/provider.tsx

### To Reproduce

boot the server and it shows in the console as an error.

### Describe the Bug

Saying `localStorage is not defined` but its a client component meaning that it should only ever be run on the client side, where this would be defined. does the same for window.

### Expected Behavior

Shouldnt be an error because it is in a client component not a server component

### Which browser are you using? (if relevant)

Chome

### How are you deploying your application? (if relevant)

Vercel

**Comment by PatrikTheDev** [comment]
I might be mistaken but "Client components" (in RSC terms) still get the standard SSR treatment, the main difference is that they get hydrated on the client (unlike Server components which essentially stay as static HTML). If you want to fully bypass SSR, use `next/dynamic` and set `{ ssr: false }` in the `dynamic` call. Hope this helps!

**Comment by TillHeinzel** [comment]
I ran into this problem, too. For now, I have solved it like this, with a custom hook:

``` TSX
function useLocallyStoredState<Type>(
  localStoreName: string,
  dummy: Type,
  defaultVal: Type
): [() => Type, (s: Type) => void] {
  const [val, setVal] = useState(dummy);

  const getVal = () => {
    if (val === dummy) {
      const load = () => {
        let data = window.localStorage.getItem(localStoreName);
        if (data === null) return defaultVal;

        const parsedData = JSON.parse(data);
        if (!parsedData) return defaultVal;

        return parsedData;
      };

      const intermediate = load();
      setVal(intermediate);
      return intermediate;
    }
    return val;
  };

  const updateVal = (newVal: Type) => {
    window.localStorage.setItem(localStoreName, JSON.stringify(newVal));
    setVal(newVal);
  };

  return [getVal, updateVal];
}
```

usage:

``` TSX
  const [getTitle, updateTitle] = useLocallyStoredState(
    "math_website_build_set_title",
    "___Dummy___",
    ""
  );
```

This works because the hooks only run client-side, while initialization code apparently runs server-side. You can get away with less if you only need strings.

It seems like an ugly, roundabout way just to load an initial state from localstorage, though. I welcome any suggestions for improvement.

**Comment by The-Code-Monkey** [comment]
The problem I have is that I'm using it to read a theme value from local
storage so I get a flash on the users side. And I also can't find a way to
grab the users preferred colour scheme on initial load

**Comment by TillHeinzel** [comment]
Yes, that flash is annoying. I am not certain how to solve it. I think, though, if your users have a preferred color scheme, maybe that should be stored with user data in a database? Then it would be accessible from the server across devices, and you would not run into this trouble with localstorage.

**Comment by boatilus** [comment]
The same can also be accomplished with a cookie, at the cost of sending a few extra bytes with each request.

**Comment by timrae** [comment]
I'm facing the same problem. I'm using the package `react-oidc-context` to handle authorization with an API. I just have a single route currently, with `page.tsx` looking as follows:

```typescript
export default function Home() {
  return (
    <AuthProvider {...oidcConfig}>
      <App />
    </AuthProvider>
  );
}
```

When I try to add `userStore: new WebStorageStateStore({ store: window.localStorage }),` to the `oidcConfig` object I get an error:

```
error - app/page.tsx (36:47) @ window
error - ReferenceError: window is not defined
```

I was also confused to see this error despite setting`"use client";` and `export const dynamic = "force-dynamic";` . If this can't be handled automatically, maybe something could be added to the docs at least.

**Comment by pedrofrxncx** [comment]
got the same issue here, the app doesn't break but it is annoying

**Comment by sharpsteelsoftware** [comment]
Yup this is driving me crazy. I have "use client" code that contains useLocalStorage() and I get this freaking error in my terminal dump that useLocalStorage is not defined. Protecting the useLocalStorage() call with a window undefined check only causes a different error: hydration of client does not match that at server time.

**Comment by pedrofrxncx** [comment]
> got the same issue here, the app doesn't break but it is annoying

actually, it does break the build process. what can I do here? oimoi

**Comment by aliblackeye** [comment]
> Yup this is driving me crazy. I have "use client" code that contains useLocalStorage() and I get this freaking error in my terminal dump that useLocalStorage is not defined. Protecting the useLocalStorage() call with a window undefined check only causes a different error: hydration of client does not match that at server time.

did you solve it?
